src/pad.rs

```diff
@@ -0,0 +1,363 @@
+                return Err(());
+            }
+
+            let digest_alg = self.digest_alg;
+            let decoded_digest =
+                try!(decoded.skip_and_get_input(digest_alg.output_len));
+            let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
+            if decoded_digest != digest.as_ref() {
+                return Err(());
+            }
+            Ok(())
+        })
+    }
+
+
+}
+
+// Maximum supported length of the salt;
+const MAX_SLEN: usize = 512 / 8;
+
+// Maxmimum supported length for intermediate value used in PSS padding.
+const MAX_MLEN: usize = 512;
+
+pub struct PSSPadding {
+    pub digest_alg: &'static digest::Algorithm,
+    pub mgf: fn(seed: &[u8], mask_len: usize, mask: &mut [u8],
+                digest_alg: &'static digest::Algorithm) -> Result<(), ()>,
+    pub s_len: usize
+}
+
+impl PaddingAlgorithm for PSSPadding {
+    // Implement padding procedure per EMSA-PSS,
+    // https://tools.ietf.org/html/rfc3447#section-9.1.
+    fn encode(&self, msg: &[u8], out: &mut [u8], rng: &rand::SecureRandom)
+              -> Result<(), ()> {
+
+        //    Steps:
+
+        //    1.  If the length of M is greater than the input limitation for the
+        //        hash function (2^61 - 1 octets for SHA-1), output "message too
+        //        long" and stop.
+
+        //    2.  Let mHash = Hash(M), an octet string of length hLen.
+
+        //    3.  If emLen < hLen + sLen + 2, output "encoding error" and stop.
+
+        let digest_len = self.digest_alg.output_len;
+        let em_len = out.len();
+        let min_len = digest_len + self.s_len + 2;
+
+        debug_assert!(em_len <= MAX_MLEN);
+        debug_assert!(self.s_len <= MAX_SLEN);
+
+        if em_len < min_len {
+            return Err(());
+        }
+        let m_hash = digest::digest(self.digest_alg, msg);
+
+        //    4.  Generate a random octet string salt of length sLen; if sLen = 0,
+        //        then salt is the empty string.
+
+        let mut salt = [0u8; MAX_SLEN];
+        try!(rng.fill(&mut salt[..self.s_len]));
+
+
+        //    5.  Let
+        //          M' = (0x)00 00 00 00 00 00 00 00 || mHash || salt;
+
+        //        M' is an octet string of length 8 + hLen + sLen with eight
+        //        initial zero octets.
+
+        //    6.  Let H = Hash(M'), an octet string of length hLen.
+
+        let mut ctx = digest::Context::new(self.digest_alg);
+
+        let prefix = [0u8; 8];
+        ctx.update(&prefix);
+        ctx.update(m_hash.as_ref());
+        ctx.update(&salt[..self.s_len]);
+
+        let h_hash = ctx.finish();
+
+        //    7.  Generate an octet string PS consisting of emLen - sLen - hLen - 2
+        //        zero octets.  The length of PS may be 0.
+
+        //    8.  Let DB = PS || 0x01 || salt; DB is an octet string of length
+        //        emLen - hLen - 1.
+
+        let mut db = [0u8; MAX_MLEN];
```

Comment:
Does db really need to be a separate array, or could it just be a slice of `out` in order to avoid the copies below?